libraries/chain/webassembly/runtimes/eos-vm-oc/code_cache.cpp

```diff
@@ -202,97 +219,185 @@ const code_descriptor* const code_cache_sync::get_descriptor_for_code_sync(const
    return &*_cache_index.push_front(std::move(std::get<code_descriptor>(result.result))).first;
 }
 
+int code_cache_base::get_huge_memfd(size_t map_size, int memfd_flags) const {
+   int ret;
+
+   ret = syscall(SYS_memfd_create, "eosvmoc_cc", MFD_CLOEXEC | memfd_flags);
+   if(ret < 0)
+      return -1;
+
+   auto close_failed_creation = fc::make_scoped_exit([&](){close(ret);});
+
+   if(ftruncate(ret, map_size) == -1)
+      return -1;
+
+   //Empirically on 5.13.9 there is no need to pass MAP_HUGETLB to have the mapping actually use hugepages.
+   //That makes sense due to the underlying implementation of memfd, but the docs seem ambiguous on expected behavior.
+
+   //The man page states w.r.t. MAP_POPULATE:
+   //     The mmap() call doesn't fail if the mapping cannot be populated (for example,
+   //     due to limitations on the number of mapped huge pages when using MAP_HUGETLB).
+   //Well, empirically on 5.13.9 MAP_POPULATE does cause a failure if there aren't actually enough hugepages available
+   // at the given size. It's not clear how to (easily) more robustly handle what the documentation suggests: where the MAP_POPULATE
+   // passively ignores that we're overcommitted. We could attempt to force the pages to be allocated in a loop after the mmap(),
+   // but we'd just be SIGBUSed if pages aren't available. Handling SIGBUS would fully resolve this I suppose, but that's more
+   // complication than I'd like.
+   //Thus there seems to be a risk according to the documentation that there may be a nasty situation where the platform indicates
+   // 1GB pages are available, MAP_POPULATE passively succeeds, so we then plan to use 1GB pages, but then the later iostreams::copy()
+   // fails via a SIGBUS because the hugepages were overcommitted and not enough are actually available. In such a situation a user
+   // would be stuck and unable to use heap/locked mode; worse, the SIGBUS may leave the main chainbase database dirty.
+   void* mapped = mmap(NULL, map_size, PROT_READ|PROT_WRITE, MAP_SHARED|MAP_POPULATE, ret, 0);
```

Comment:
It does appear from the docs that MAP_POPULATE may behave as you describe. Not sure we have much choice. Seems like a decent way forward at the moment with a look toward replacement of all of this with just serialization to/from memory on startup/shutdown.

Reply:
Even if we serialize everything on start/stop, I think we'll still want the ability to throw all of OC's code in to a huge page because it does have measurable performance improvement.

I looked around at some other implementations and I couldn't find any one being more paranoid here though. For example, my take on looking at mimalloc's implementation (a general purpose allocator that will try to map in 1GB pages) is that if the mmap() succeeds they simply run with it.

Probably catching the SIGBUS is the only reasonable way, it seems.